### PR TITLE
Support binding `std::optional<T>` implementation to `IReference<T>` parameters

### LINKED ISSUE #1006: Support binding std::optional<T> implementation to IReference<T> parameters
Assume I have this IDL:

```c#
runtimeclass MyThing {
    Windows.Foundation.IReference<UInt32> GetSize();
}
```

And my implementation says:

```c++
struct MyThing : MyThingT<MyThing> {
    std::optional<uint32_t> m_sizeComputed;
    winrt::Windows::Foundation::IReference<uint32_t> GetSize() {
        if (!m_sizeComputed) {
            return nullptr;
        } else {
            return winrt::box_value(m_sizeComputed.value()).as<winrt::Windows::Foundation<IReference<uint32_t>>>();
    }
}
```

I'd like this to be easier on the implementation, where instead I can just say:

```c++
struct MyThing : MyThingT<MyThing> {
    std::optional<uint32_t> m_sizeComputed;
    const std::optional<uint32_t> GetSize() { return m_sizeComputed; }
}
```

Is there a way for C++/WinRT's shims to match up `IReference<T>`-things (properties, returned types, parameters) and auto-convert them to/from `std::optional<T>`?

### PR DESCRIPTION
**Pull request by kennykerr**
Fixes #1006

### CODE REVIEW

**Comment by kennykerr** [comment]
@jonwis @sylveon looks good?

File: test/test/optional.cpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+TEST_CASE("optional")
+{
+    test_component::Optional object;
+
+    object.Property(Windows::Foundation::IReference(123));
+    REQUIRE(object.Property().Value() == 123);
+
+    object.Property(nullptr);
+    REQUIRE(object.Property() == nullptr);
```
Comment (oldnewthing) [inline test/test/optional.cpp:14]: Also

```cpp
REQUIRE(object.Property() == std::optional<int32_t>());
REQUIRE(object.Property() == std::nullopt);
```

to verify that implicit conversion to `std::optional<int32_t>` can successfully compare with std::nullopt.

**Review by oldnewthing — APPROVED** [review]

File: test/test/optional.cpp (hunk 1 above)
Comment (kennykerr) [reply test/test/optional.cpp:14]: I don't think C++ supports this. The C++ compiler would have to perform an implicit `operator T` followed by an `operator ==` and that's more than C++ is willing to do implicitly. I would have to add `operator ==` for `std::optional` if this is desired.

**Comment by sylveon** [comment]
LGTM, will help with some code that did pass around optionals but then needed to pass it to WinRT.

File: test/test/optional.cpp (hunk 1 above)
Comment (oldnewthing) [reply test/test/optional.cpp:14]: Okay well then nevermind then. May call out in a comment that those operations *don't* work.

Comment (kennykerr) [reply test/test/optional.cpp:14]: Let's consider adopting C++20's spaceship operator to simplify stuff like this.

**Pull request merged**